torch/_dynamo/_trace_wrapped_higher_order_op.py

```diff
@@ -0,0 +1,120 @@
+import functools
+
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey
+from torch._higher_order_ops.utils import autograd_not_implemented
+
+from torch._ops import HigherOrderOperator
+from torch._subclasses import FakeTensorMode
+
+from torch.fx.experimental.proxy_tensor import ProxyTorchDispatchMode, track_tensor_tree
+from torch.utils._python_dispatch import _get_current_dispatch_mode
+
+
+__all__ = ["trace_wrapped"]
+
+
+# Trace wrapped is a higher order op meant for both invoking a bound function,
+# and for registering it as a call_function in the backward graph.
+# This allows us to re-enter dynamo during compiled autograd to trace (or graph break)
+# the functions as needed. This, in turn, means we can support functions in backward with complex python
+# state mutation. If we were to not do this, the functions would get inlined into their composing aten ops,
+# and we would lose the python state mutation.
+def trace_wrapped(*args, fn):
+    return _trace_wrapped_op(*args, fn=fn)
+
+
+_trace_wrapped_op = HigherOrderOperator("trace_wrapped")
+
+
+def _self_invoke(*args, fn):
+    # This wrapper intercepts calls to fn, and calls the real fn via _trace_wrapped
+    # Dynamo unpacks this higher order op into calling the wrapped fn
+    return _trace_wrapped_op(*args, fn=fn)
+
+
+def _assert_meta(grad, size, stride, dtype):
```

Comment:
noob dynamo q: how does dynamo know to execute these asserts at compile time (while dynamo is tracing), instead of automatically trying to add these asserts and metadata calls as proxies into the backward graph?